pkg/controllers/nodeipam/nodeipam_controller.go

```diff
@@ -0,0 +1,218 @@
+	kubeClient clientset.Interface,
+	cloud cloudprovider.Interface,
+	nodeMonitorPeriod time.Duration,
+	nodeIpamConfig config.NodeIPAMControllerConfiguration,
+) (*Controller, error) {
+	var err error
+	awsCloud, ok := cloud.(*awsv1.Cloud)
+	if !ok {
+		err = fmt.Errorf("nodeipam controller does not support %v provider", cloud.ProviderName())
+		return nil, err
+	}
+
+	var rateLimiter workqueue.RateLimiter
+	var rateLimitEnabled bool
+	if nodeIpamConfig.RateLimit > 0.0 && nodeIpamConfig.BurstLimit > 0 {
+		klog.Infof("Rate limit enabled on controller with rate %f and burst %d.", nodeIpamConfig.RateLimit, nodeIpamConfig.BurstLimit)
+		// This is the workqueue.DefaultControllerRateLimiter() but in case where throttling is enabled on the controller,
+		// the rate and burst values are set to the provided values.
+		rateLimiter = workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second),
+			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(nodeIpamConfig.RateLimit), nodeIpamConfig.BurstLimit)},
+		)
+		rateLimitEnabled = true
+	} else {
+		klog.Infof("Rate limit disabled on controller.")
+		rateLimiter = workqueue.DefaultControllerRateLimiter()
+		rateLimitEnabled = false
+	}
+
+	nc := &Controller{
+		nodeInformer:      nodeInformer,
+		kubeClient:        kubeClient,
+		cloud:             awsCloud,
+		workqueue:         workqueue.NewNamedRateLimitingQueue(rateLimiter, "NodeIpam"),
+		nodesSynced:       nodeInformer.Informer().HasSynced,
+		nodeMonitorPeriod: nodeMonitorPeriod,
+		rateLimitEnabled:  rateLimitEnabled,
+	}
+
+	// for IPv6 only
+	if !nodeIpamConfig.DualStack {
+		ipam.RegisterMetrics()
+		nc.ipv6CIDRAllocator, err = ipam.NewIPv6RangeAllocator(kubeClient, nodeInformer, awsCloud, rateLimiter, rateLimitEnabled, nodeMonitorPeriod)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		eventBroadcaster := record.NewBroadcaster()
+		recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "cidrAllocator"})
+		eventBroadcaster.StartStructuredLogging(0)
+		klog.V(0).Infof("Sending events to api server.")
+		eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+		allocatorParams := ipam.CIDRAllocatorParams{
+			ClusterCIDRs:      nodeIpamConfig.ClusterCIDRs,
+			NodeCIDRMaskSizes: []int{int(nodeIpamConfig.NodeCIDRMaskSize)},
+		}
+		nc = &Controller{
+			kubeClient:            kubeClient,
+			clusterCIDRs:          allocatorParams.ClusterCIDRs,
+			cloud:                 awsCloud,
+			cidrSets:              []*cidrset.CidrSet{},
+			nodeLister:            nodeInformer.Lister(),
+			nodesSynced:           nodeInformer.Informer().HasSynced,
+			nodeCIDRUpdateChannel: make(chan ipam.NodeReservedCIDRs, ipam.CidrUpdateQueueSize),
+			recorder:              recorder,
+			nodesInProcessing:     sets.NewString(),
+		}
+		nodeList, err := listNodes(kubeClient)
```

Comment:
why we need to use listNodes? shouldn't the informer handles all existing nodes?